src/migrate.c

```diff
@@ -0,0 +1,171 @@
+#include <string.h>
+
+#include "redisraft.h"
+
+static int validSlotMagic(RedisRaftCtx *rr, unsigned int slot, int magic)
+{
+    /* validSlot will have already validated that slot is importing */
+    ShardGroup *sg = rr->sharding_info->importing_slots_map[slot];
+
+    for (unsigned int i = 0; i < sg->slot_ranges_num; i++) {
+        ShardGroupSlotRange *sr = &sg->slot_ranges[i];
+        if (sr->start_slot <= slot && sr->end_slot >= slot && sr->magic == magic) {
+            return 1;
+        }
+    }
+
+    return 0;
+}
+
+static int validSlotTerm(RedisRaftCtx *rr, int slot, raft_term_t term)
+{
+    ShardingInfo *si = rr->sharding_info;
+
+    if (si->max_importing_term[slot] <= term) {
+        si->max_importing_term[slot] = term;
+        return 1;
+    }
+
+    return 0;
+}
+
+static int validSlot(RedisRaftCtx *rr, int slot)
+{
+    ShardGroup *sg = rr->sharding_info->importing_slots_map[slot];
+
+    if (sg && sg->local) {
+        return 1;
+    }
+
+    return 0;
+}
+
+void importKeys(RedisRaftCtx *rr, raft_entry_t *entry)
+{
+    RedisModule_Assert(entry->type == RAFT_LOGTYPE_IMPORT_KEYS);
+
+    RaftReq *req = entry->user_data;
+
+    ImportKeys import_keys = {0};
+    RedisModule_Assert(RaftRedisDeserializeImport(&import_keys, entry->data, entry->data_len) == RR_OK);
+    RedisModule_Assert(import_keys.num_keys > 0);
+
+    // FIXME: validate no cross slot migration at append time
+    int slot = KeyHashSlotRedisString(import_keys.key_names[0]);
+
+    if (!validSlot(rr, slot)) {
+        if (req) {
+            RedisModule_ReplyWithError(req->ctx, "ERR not an importing slot");
+        }
+        goto exit;
+    }
+
+    if (!validSlotMagic(rr, slot, import_keys.magic)) {
+        if (req) {
+            RedisModule_ReplyWithError(req->ctx, "ERR invalid magic");
+        }
+        goto exit;
+    }
+
+    if (!validSlotTerm(rr, slot, import_keys.term)) {
+        // this is acceptable (suspended old leader
+        LOG_WARNING("ignoring import keys as old term");
```

Comment:
I think this kind of logging is problematic as it can blow up logs very quickly:
* Consider making this a debug message.
* Consider having a counter for this, i.e. count the number of import keys ignored for different reasons. Note that we could try to leverage the fact that Redis is tracking different errors (`INFO errorstats`).

Reply:
documented as https://github.com/RedisLabs/redisraft/issues/357